build/launch_training.py

```diff
@@ -53,78 +61,124 @@ def main():
 
     logging.info("Initializing launch training script")
 
-    job_config = get_job_config()
-
-    logging.debug("Input params parsed: %s", job_config)
-
-    (
-        model_args,
-        data_args,
-        training_args,
-        tune_config,
-        merge_model,
-        file_logger_config,
-        aim_config,
-    ) = process_launch_training_args(job_config)
+    try:
+        job_config = get_job_config()
+        logging.debug("Input params parsed: %s", job_config)
+
+        (
+            model_args,
+            data_args,
+            training_args,
+            tune_config,
+            merge_model,
+            file_logger_config,
+            aim_config,
+        ) = process_launch_training_args(job_config)
+    except Exception as e:  # pylint: disable=broad-except
+        logging.error(traceback.format_exc())
+        write_termination_log(
+            "Exception raised during training. This may be a problem with your input: {}".format(
+                e
+            )
+        )
+        sys.exit(USER_ERROR_EXIT_CODE)
 
     original_output_dir = training_args.output_dir
     with tempfile.TemporaryDirectory() as tempdir:
         training_args.output_dir = tempdir
-        tracker_config_args = TrackerConfigFactory(
-            file_logger_config=file_logger_config, aim_config=aim_config
-        )
-        sft_trainer.train(
-            model_args=model_args,
-            data_args=data_args,
-            train_args=training_args,
-            peft_config=tune_config,
-            tracker_configs=tracker_config_args,
-        )
-
-        if merge_model:
-            export_path = os.getenv(
-                "LORA_MERGE_MODELS_EXPORT_PATH", original_output_dir
+        try:
+            tracker_config_args = TrackerConfigFactory(
+                file_logger_config=file_logger_config, aim_config=aim_config
             )
-
-            # get the highest checkpoint dir (last checkpoint)
-            lora_checkpoint_dir = get_highest_checkpoint(training_args.output_dir)
-            full_checkpoint_dir = os.path.join(
-                training_args.output_dir, lora_checkpoint_dir
+            sft_trainer.train(
+                model_args=model_args,
+                data_args=data_args,
+                train_args=training_args,
+                peft_config=tune_config,
+                tracker_configs=tracker_config_args,
             )
-
-            logging.info(
-                "Merging lora tuned checkpoint %s with base model into output path: %s",
-                lora_checkpoint_dir,
-                export_path,
+        except MemoryError:
```

Comment:
The error that i've seen a lot is `torch.cuda.OutOfMemoryError`, which I think is a different error than this one. If so, we should add it in. The error I've seen a lot is:
```
torch.cuda.OutOfMemoryError: CUDA out of memory. Tried to allocate 524.00 MiB. GPU 0 has a total capacity of 79.15 GiB of which 176.38 MiB is free. Process 360061 has 78.98 GiB memory in use. Of the allocated memory 77.09 GiB is allocated by PyTorch, and 403.78 MiB is reserved by PyTorch but unallocated. If reserved but unallocated memory is large try setting PYTORCH_CUDA_ALLOC_CONF=expandable_segments:True to avoid fragmentation.  See documentation for Memory Management  (https://pytorch.org/docs/stable/notes/cuda.html#environment-variables)
```

Reply:
The other memory issue I've seen is with ephemeral storage but I don't know if we can catch it. It usually causes a problem and the pod is killed by Kubernetes. This is what the pod details show afterwards:
```yaml
containerStatuses:
  - containerID: cri-o://eee03dd0662a24fd081d0445cc5006d9c941d546ec260b1bc336737e2b4958d4
    image: docker-na-public.artifactory.swg-devops.com/wcp-ai-foundation-team-docker-virtual/sft-trainer:7f32893_ubi9_py311
    imageID: docker-na-public.artifactory.swg-devops.com/wcp-ai-foundation-team-docker-virtual/sft-trainer@sha256:4f64510f4c0cc19ec7e00558b26ed30117528e3aa7977761ee69061e82c39a72
    lastState: {}
    name: train-conductor-training
    ready: false
    restartCount: 0
    started: false
    state:
      terminated:
        containerID: cri-o://eee03dd0662a24fd081d0445cc5006d9c941d546ec260b1bc336737e2b4958d4
        exitCode: 1
        finishedAt: "2024-04-12T19:35:26Z"
        reason: Error
        startedAt: "2024-04-12T18:09:39Z"
  hostIP: 10.0.0.60
  message: 'The node was low on resource: ephemeral-storage. '
  phase: Failed
  podIP: 10.135.5.133
  podIPs:
  - ip: 10.135.5.133
  qosClass: Burstable
  reason: Evicted
  startTime: "2024-04-12T18:09:34Z"
```

Reply:
> The error that i've seen a lot is `torch.cuda.OutOfMemoryError`, which I think is a different error than this one. If so, we should add it in. The error I've seen a lot is:
> 
> ```
> torch.cuda.OutOfMemoryError: CUDA out of memory. Tried to allocate 524.00 MiB. GPU 0 has a total capacity of 79.15 GiB of which 176.38 MiB is free. Process 360061 has 78.98 GiB memory in use. Of the allocated memory 77.09 GiB is allocated by PyTorch, and 403.78 MiB is reserved by PyTorch but unallocated. If reserved but unallocated memory is large try setting PYTORCH_CUDA_ALLOC_CONF=expandable_segments:True to avoid fragmentation.  See documentation for Memory Management  (https://pytorch.org/docs/stable/notes/cuda.html#environment-variables)
> ```

@anhuong I've also observed a `RuntimeError` exception with the text `CUDA error: out of memory`.

Reply:
Great, thanks. I added a specific catch in for `torch.cuda.OutOfMemoryError`. I was able to catch one when deploying a large model on a resource constrained server :) 
```
Containers:
  pytorch:
    Container ID:   xxxx
    Image:         xxx
    Image ID:       xxx
    Port:           xxx
    Host Port:      0/TCP
    State:          Terminated
      Reason:       Error
      Message:      OOM error during training. CUDA out of memory. Tried to allocate 172.00 MiB. GPU
      Exit Code:    203
```

Reply:
It's not clear to me that we could catch the ephemeral-storage problem in code